File: leetcode/algorithms/go/0011-container-with-most-water.go
package leetcode

func maxArea(height []int) int {
	var result, i, j = 0, 0, len(height) - 1
	for i < j {
		h := height[i]
		if height[j] < h {
			h = height[j]
		}
		if result < h*(j-i) {
			result = h * (j - i)
		}
		if h == height[i] {
			i++
		} else {
			j--
		}
	}
	return result
}

// func maxArea(height []int) int {
// 	max := func(a, b int) int {
// 		if a > b {
// 			return a
// 		}
// 		return b
// 	}
// 	min := func(a, b int) int {
// 		if a < b {
// 			return a
// 		}
// 		return b
// 	}
// 	result, i, j := 0, 0, len(height)-1
// 	for i < j {
// 		result = max(result, (min(height[i], height[j]) * (j - i)))
// 		if height[i] < height[j] {
// 			i++
// 		} else {
// 			j--
// 		}
// 	}
// 	return result
// }

// math.Max()
// math.Min()

File: leetcode/algorithms/javascript/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
    let result = 0;
    let i = 0;
    let j = height.length - 1;
    while (i < j) {
        let h = height[i];
        if (height[j] < h) {
            h = height[j];
        }
        if (result < h * (j - i)) {
            result = h * (j - i);
        }
        if (h == height[i]) {
            i++;
        } else {
            j--;
        }
    }
    return result;
};

// var maxArea = function (height) {
//     let result = 0;
//     let i = 0, j = height.length - 1;
//     while (i < j) {
//         result = Math.max(result, Math.min(height[i], height[j]) * (j - i));
//         if (height[i] < height[j]) {
//             i++;
//         } else {
//             j--;
//         }
//     }
//     return result;
// };
